lib/datadog/core/telemetry/metric_queue.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+module Datadog
+  module Core
+    module Telemetry
+      # Stores all the metrics information by request_type during a time interval
+      class MetricQueue
+        def initialize
+          @metrics = {
+            'generate-metrics' => {},
+            'distributions' => {},
+          }
+        end
+
+        def add_metric(namespace, name, value, tags, metric_klass)
+          namespace_space = @metrics[metric_klass.request_type][namespace] ||= {}
+          existing_metric = namespace_space[name]
+
+          if existing_metric
+            existing_metric.update_value(value)
+            @metrics[metric_klass.request_type][namespace][name] = existing_metric
+          else
+            new_metric = metric_klass.new(name, tags)
+            new_metric.update_value(value)
+            @metrics[metric_klass.request_type][namespace][name] = new_metric
+          end
+        end
+
+        def build_metrics_payload
```

Comment:
Do we have to reset the metric counters when we flush?

Reply:
Yes. 

We do in the client right after sending the metrics. 
https://github.com/DataDog/dd-trace-rb/blob/27142bdee303ad61fe395fe43c3f22624b3b5b6c/lib/datadog/core/telemetry/client.rb#L120-L124